hazelcast/src/main/java/com/hazelcast/instance/impl/DefaultAddressPicker.java

```diff
@@ -381,30 +380,53 @@ AddressDefinition pickMatchingAddress(Collection<InterfaceDefinition> interfaces
                 if (address == null) {
                     continue;
                 }
-                matchingAddress = address;
-
-                if (preferIPv6Addresses) {
-                    // IPv6 address is preferred, return if address is IPv6.
-                    if (inetAddress instanceof Inet6Address) {
-                        return matchingAddress;
-                    }
-                } else if (inetAddress instanceof Inet4Address) {
-                    // No IPv6 address preference, return if address is IPv4.
-                    return matchingAddress;
+                if (getPriority(address, preferIPv6Addresses) > getPriority(matchingAddress, preferIPv6Addresses)) {
+                    matchingAddress = address;
                 }
             }
         }
         // nothing matched to IP version preference, return what we have.
```

Comment:
Biggest nit-pick ever: This comment isn't accurate anymore, since the intention is to return `matchingAddress` at the end of the loop (address with highest priority)